### PR TITLE
[SPARK-26032][PYTHON] Break large sql/tests.py files into smaller files

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This is the official first attempt to break huge single `tests.py` file - I did it locally before few times and gave up for some reasons. Now, currently it really makes the unittests super hard to read and difficult to check. To me, it even bothers me to to scroll down the big file. It's one single 7000 lines file!

This is not only readability issue. Since one big test takes most of tests time, the tests don't run in parallel fully - although it will costs to start and stop the context.

We could pick up one example and follow. Given my investigation, the current style looks closer to NumPy structure and looks easier to follow. Please see https://github.com/numpy/numpy/tree/master/numpy.

Basically this PR proposes to break down `pyspark/sql/tests.py` into ...:

```bash
pyspark
...
├── sql
...
│   ├── tests  # Includes all tests broken down from 'pyspark/sql/tests.py'
│   │   │      # Each matchs to module in 'pyspark/sql'. Additionally, some logical group can
│   │   │      # be added. For instance, 'test_arrow.py', 'test_datasources.py' ...
│   │   ├── __init__.py
│   │   ├── test_appsubmit.py
│   │   ├── test_arrow.py
│   │   ├── test_catalog.py
│   │   ├── test_column.py
│   │   ├── test_conf.py
│   │   ├── test_context.py
│   │   ├── test_dataframe.py
│   │   ├── test_datasources.py
│   │   ├── test_functions.py
│   │   ├── test_group.py
│   │   ├── test_pandas_udf.py
│   │   ├── test_pandas_udf_grouped_agg.py
│   │   ├── test_pandas_udf_grouped_map.py
│   │   ├── test_pandas_udf_scalar.py
│   │   ├── test_pandas_udf_window.py
│   │   ├── test_readwriter.py
│   │   ├── test_serde.py
│   │   ├── test_session.py
│   │   ├── test_streaming.py
│   │   ├── test_types.py
│   │   ├── test_udf.py
│   │   └── test_utils.py
...
├── testing  # Includes testing utils that can be used in unittests.
│   ├── __init__.py
│   └── sqlutils.py
...
```

## How was this patch tested?

Existing tests should cover.

`cd python` and `./run-tests-with-coverage`. Manually checked they are actually being ran.

Each test (not officially) can be ran via:

```
SPARK_TESTING=1 ./bin/pyspark pyspark.sql.tests.test_pandas_udf_scalar
```

Note that if you're using Mac and Python 3, you might have to `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES`.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@BryanCutler and @squito, Here is the official first attempt to break `pyspark/sql/tests.py` into multiple small files.

If there are no outstanding issues (for instance, if we are agreed upon the high level approach), I would like to push this quicker as this will super likely have conflicts easily.

**Comment by HyukjinKwon** [comment]
adding @rxin (derived from mailing list)

**Comment by HyukjinKwon** [comment]
FWIW, I at least double checked if they are any tests missing, and if they are actually being ran (via coverage).

**Comment by HyukjinKwon** [comment]
For your information, here's the line counts for each file:

```
      52 ./test_utils.py
     197 ./test_catalog.py
      43 ./test_group.py
     318 ./test_session.py
     151 ./test_readwriter.py
    2180 ./test_pandas_udf.py
     564 ./test_streaming.py
      53 ./test_conf.py
     155 ./test_column.py
     652 ./test_udf.py
     276 ./test_functions.py
     261 ./test_context.py
     136 ./test_serde.py
     168 ./test_datasources.py
     397 ./test_arrow.py
      94 ./test_appsubmit.py
     942 ./test_types.py
     735 ./test_dataframe.py
```

It's rather evenly distributed, except for `./test_pandas_udf.py`. Maybe we should make `test_scalar_pandas_udf.py` and etc. but I would like to avoid this for now.

**Comment by HyukjinKwon** [comment]
Elapsed time looks virtually same. All tests looks running fine. The last commit should show skipped tests fine as well. Should be ready for a look.

**Review by dongjoon-hyun — COMMENTED** [review]
It's cool! @HyukjinKwon .
Could you add some descriptions to run a single test file or a single test case if exists?

**Comment by squito** [comment]
thanks for working on this @HyukjinKwon .  Did you test on python3 as well?

I'm surprised elapsed time is the same -- I would have expected it to be faster as the tests could run in parallel.

aside: does anybody have a good recommendation for tools to help with viewing a diff like this?  I feel like I want something really simple: a tool to show me where each original test got moved to, and highlight what was deleted or added.  I tried p4merge and beyond compare, but neither was helpful (or I'm using them wrong).

**Comment by rxin** [comment]
Great initiative!

I'd break the pandas udf one into smaller pieces too, as you suggested. We should also investigate why the runtime didn't improve ...

File: python/pyspark/testing/sqlutils.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
+#
```
Comment (BryanCutler) [inline python/pyspark/testing/sqlutils.py]: Maybe rename this file to `sql_testing_utils.py`? And is it possible to just place this in the `sql/tests` directory? Having the `testing` folder seems a little ambiguous to me

**Review by BryanCutler — COMMENTED** [review]
Thanks for doing this @HyukjinKwon! +1 on splitting up the pandas_udf tests, but that can be done later. I ran some tests locally and looks good. I'm just wondering what you think about putting the `sqlutils.py` under `sql/tests` since that is where it is being used.

**Comment by HyukjinKwon** [comment]
Yup, will break pandas one into smaller ones as well.

File: python/pyspark/testing/sqlutils.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/testing/sqlutils.py]: Yea, similar thought. One thing is though testing util itself has some dependencies to each others (for instance, `from pyspark.tests import ReusedPySparkTestCase`). So, I thought it's okay to make it separate.

Another benefit from doing so is we can simply refer the example (NumPy) as is. As you already know, NumPy has been there more then 15 years and should be mature enough. It should be okay to trust their approach.

**Comment by HyukjinKwon** [comment]
> I'd break the pandas udf one into smaller pieces too, as you suggested. We should also investigate why the runtime didn't improve ...

One suspection from my investigation is, it requires to stop and start the context for each test which is costly. I also expected it's going to slightly decrease the time but actually it looks that slightly increased the time (I guess 2 ~ 3 mins in total? - shouldn't be a big deal).

**Comment by HyukjinKwon** [comment]
I am going to push after testing and double checking. The line counts would look like this

```
      54 ./test_utils.py
     199 ./test_catalog.py
     503 ./test_grouped_agg_pandas_udf.py
      45 ./test_group.py
     320 ./test_session.py
     153 ./test_readwriter.py
     806 ./test_scalar_pandas_udf.py
     216 ./test_pandas_udf.py
     566 ./test_streaming.py
      55 ./test_conf.py
      16 ./__init__.py
     530 ./test_grouped_map_pandas_udf.py
     157 ./test_column.py
     654 ./test_udf.py
     262 ./test_window_pandas_udf.py
     278 ./test_functions.py
     263 ./test_context.py
     138 ./test_serde.py
     170 ./test_datasources.py
     399 ./test_arrow.py
      96 ./test_appsubmit.py
     944 ./test_types.py
     737 ./test_dataframe.py
```

**Comment by rxin** [comment]
One thing - I would put “pandas” right after test_ so you get the natural
logical grouping with sorting by file name.

**Comment by HyukjinKwon** [comment]
Yup!

**Comment by HyukjinKwon** [comment]
> Could you add some descriptions to run a single test file or a single test case if exists?

Done!

**Comment by HyukjinKwon** [comment]
> Did you test on python3 as well?

Of course!

**Comment by HyukjinKwon** [comment]
I am merging this in - maybe I am rushing it but please allow me to go ahead since it's going to block other PySpark PRs.

At worst case, I am willing to revert and propose this again if there's something obviously wrong. Please feel free to leave some comments even after this is merged. I would appreciate it.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**